lib/portlayer/storage/storage.go

```diff
@@ -12,6 +12,74 @@
 
+// Package storage provides abstractions for interacting with storage locations, both management and access.
+// There are two primary type of storage handled by this package at this time:
+// a. VMDKs on a vSphere datastore
+// b. NFS shares
+//
+// We have more than two sub-packages however as we use distinct packages to provide different lifecycle and
+// usage semantics for types of data using those mechanisms. Examples include:
+// * application logic (images)
+// * application transient data (image read/write layer)
+// * application persistent data (volumes)
+//
+// As such there are a set of concepts captured as interfaces:
+// * store
+// * resolver
+// * exporter
+// * importer
+// * data source
+// * data sink
+//
+// The odd one out for these concepts is the `store`. This is because we did not want to be prescriptive about
+// what a store supports, but it may well distill into an interface at a later date. Currently the various
+// store implementations consist of a composition of `resolver`, `importer`, and `exporter` interfaces.
+//
+// The `data source` and `data sink` interfaces are present to allow abstracted read and write access respectively
+// _to a specific storage mechanism. The data source and data sink instances should be considered single use and not
+// reused for multiple access. This semantic was chosen as it's the least prescriptive on storage availability and
+// connectivity - the precise means used to read or write from storage may change, for example, based on whether
+// the storage is in use (has an owner).
+// The `resolver` interface is present to translate a potentially store specific ID into an unambiguous URI, and to
+// allow discovery of which virtual machines, if any, currently claim that ID. This returns virtual machines instead
+// of a further abstracted interface because it's not clear at time of writing what that abstraction should or would
+// be.
+// The `importer` and `exporter` interfaces are the factories used to acquire `data source` and `data sink` instances.
+// They also provide convenience methods for directly creating and consuming the source/sink in one operation.
+//
+// It is not required that all storage elements implement both Importer and Exporter mechanisms - having write only or
+// read only storage is viable.
+//
+// The data source and sink interfaces also specify a basic accessor method for the underlying mechanism. This was
+// added to allow for some future-proofing, with the primary extension expected to be a generic FileWalker interface
+// so that directory walk and listings can be done without needing to care about the access mechanism (e.g. common
+// between a local filesystem mount, an XDR NFS client, or a VM toolbox client).
+//
+// The general structure of a specific storage implementation is as follows and is imposed so that it is easier to
+// locate a specific functional path for a given usage/storage type combination:
+// ```
+// storage/ - this package
```

Comment:
I like the inclusion of a file tree here.

Some thoughts:
1. It'd be good to include `storage.go`, at least to explain when something should go in `storage.go` vs. one of the other top-level files — something simple like "interfaces, `RegisterImporter`, and `RegisterExporter`" may encourage people to put other stuff in a more specific file.
2. It seems like `errors.go` is a consistent convention for the existing code, and is probably worth including.
3. It seems like `store.go` is a consistent convention for the existing code, and is probably worth including.
4. I found it a little hard to differentiate between "variables" (e.g., `implementation`) and "literals" (e.g., `cache`).

In order words, how about something like:

```
// storage/                  - this package
// ├── [type]/               - the high level type of use, implies semantics (e.g., volume)
// │   ├── [type].go         - the common store implementation aspects across implementations
// │   ├── errors.go         - ???
// │   ├── cache             - a `store` compliant cache implementation, if needed, with appropriate semantics for the type of use
// │   │   └── cache.go
// │   └── [implementation]/ - a specific implementation type (e.g., vsphere)
// │       ├── bind.go       - modifies a portlayer handle to configure active use of a specific `joined` storage
// │       ├── export.go     - implements the `read` side of the interfaces (Exporter and DataSource)
// │       ├── import.go     - a specific implementation type
// │       ├── join.go       - modifies a portlayer handle to configure basic access to a specific instance of the storage type
// │       └── store.go      - ???
// ├── storage.go            - ???
// └── [purpose].go          - common functions used by the various type/implementation combinations
```